packages/apollo-server-core/src/utils/schemaManager.ts

```diff
@@ -0,0 +1,216 @@
+   * - Initialize schema-derived data.
+   * - Synchronously notify onSchemaLoadOrUpdate() listeners of schema load, and
+   *   asynchronously notify them of schema updates.
+   */
+  public async start(): Promise<void> {
+    if (this.modeSpecificState.mode === 'gateway') {
+      const gateway = this.modeSpecificState.gateway;
+      if (gateway.onSchemaLoadOrUpdate) {
+        // Use onSchemaLoadOrUpdate if available, as it reports the core
+        // supergraph SDL and always reports the initial schema load.
+        this.modeSpecificState.unsubscribeFromGateway =
+          gateway.onSchemaLoadOrUpdate((schemaContext) => {
+            this.processSchemaLoadOrUpdateEvent(schemaContext);
+          });
+      } else {
+        this.modeSpecificState.unsubscribeFromGateway = gateway.onSchemaChange(
+          (apiSchema) => {
+            this.processSchemaLoadOrUpdateEvent({ apiSchema });
+          },
+        );
+      }
+
+      const config = await this.modeSpecificState.gateway.load({
+        apollo: this.modeSpecificState.apolloConfig,
+      });
+
+      // Note that for old gateways that have onSchemaChange() and no
+      // onSchemaLoadOrUpdate(), this.schemaDerivedData may not be initialized
+      // during gateway.load() (because old gateways don't notify listeners on
+      // schema load in some cases), so we must initialize it here if needed.
+      if (!this.schemaDerivedData) {
+        this.processSchemaLoadOrUpdateEvent({ apiSchema: config.schema });
+      }
+    } else {
+      this.processSchemaLoadOrUpdateEvent({
+        apiSchema: this.modeSpecificState.apiSchema,
+      });
+    }
+  }
+
+  /**
+   * Registers a listener for schema load/update events. Note that the latest
+   * event is buffered, i.e.
+   * - If registered before start(), then the callback will be first called in
+   *   start(), and then later for updates.
+   * - If registered after start() but before stop(), the callback will be first
+   *   called in this method (for whatever the current schema is), and then
+   *   later for updates.
+   * - If registered after stop(), the callback will never be called.
+   *
+   * For gateways, a core supergraph SDL will be provided to the callback. If
+   * your gateway is too old to provide a core supergraph SDL, this method will
+   * throw.
+   *
+   * @param callback The listener to execute on schema load/updates.
+   */
+  public onSchemaLoadOrUpdate(
+    callback: (schemaContext: GraphQLSchemaContext) => void,
+  ): Unsubscriber {
+    if (
+      this.modeSpecificState.mode === 'gateway' &&
+      !this.modeSpecificState.gateway.onSchemaLoadOrUpdate
+    ) {
+      throw new Error(
+        'Your gateway is too old to register a onSchemaLoadOrUpdate() listener',
+      );
+    } else {
+      if (!this.isStopped && this.schemaContext) {
+        try {
+          callback(this.schemaContext);
+        } catch (e) {
+          this.logger.error(
```

Comment:
This method is only called during `ApolloServer._start`, so if the callback throws (eg with "The overrideReportedSchema option is incompatible with gateways") it would be reasonable to make startup fail. Let's not get back into that weird world where errors during startup are just logged instead of causing startup to fail, even for something like these plugins. (Wrapping the error in something naming the plugin hook and rethrowing sounds reasonable though.)

Logging is reasonable for errors that occur *after* system startup here, but the whole point of `ApolloServer.start()` is that it's a thing that can easily make the server fail before listening if anything goes wrong.

Reply:
Since it's only called in `_start()`, I'm down for having `onSchemaLoadOrUpdate()` catch and re-throw here. I've kept the logging as well, since it would appear JS `Error`s don't have a `cause` argument like Java `Throwable`s.

Reply:
I generally am not super into "log and rethrow" — if you're rethrowing then the caller should be able to catch and handle without there also being a log they have to suppress?

Reply:
The main reason I'm logging there is in the event they want the stacktrace of the original error thrown by the plugin (which would disappear in the rethrow). But arguably they could add a catch/log/rethrow to their plugin code if they were interested in that, so I'm also down to remove the log.